Extension/src/Utility/System/info.ts

```diff
@@ -0,0 +1,146 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { is } from './guards';
+import { AribtraryObject, Constructor } from './types';
+
+export function typeOf(instance: any) {
+    const t = typeof instance;
+    const c = (classOf(instance) as any);
+    switch (t) {
+        case 'number':
+            return Number.isNaN(instance) ? 'NaN' : 'number';
+
+        case 'object':
+            if (instance === null) {
+                return 'null';
+            }
+            if (is.promise(instance)) {
+                return c.class ? `Promise<${classOf(c.class)?.name || parentClassOf(classOf(c.class)!)?.name}>` : 'Promise';
+            }
+            return classOf(instance)?.name || parentClassOf(classOf(instance)!)?.name || '<anonymous>';
+        case 'function':
+            if (is.Constructor(instance)) {
+                return `class ${c?.name || parentClassOf(c!)?.name || '<anonymous>'}`;
+            }
+            return 'function';
+
+    }
+    return t;
+}
+
+export function hierarchy(instance: AribtraryObject | Constructor): string[] {
+    const result = new Array<string>();
+    let type = classOf(instance);
+    while (type) {
+        if (type.name) {
+            result.push(type.name);
+        }
+        type = parentClassOf(type);
+    }
+    return result;
+}
+
+export function parentClassOf(instance: AribtraryObject | Constructor): Constructor | undefined {
+    if (is.nullish(instance)) {
+        return undefined;
+    }
+    const parent = Object.getPrototypeOf(typeof instance === 'function' ? instance : instance.constructor);
+    return parent.name ? parent : undefined;
+}
+
+export function classOf(instance: AribtraryObject | Constructor): Constructor | undefined {
+    return instance ? typeof instance === 'function' ? // is it a javascript function of some kind?
+        is.asyncConstructor(instance) ? classOf(instance.class) :
+            is.Constructor(instance) ? instance as Constructor // is it really a constructor?
+                : undefined // no, it's a js function, but not a constructor

```

Comment:
Is this supposed to be JavaScript or javascript?

Reply:
fixed